### PR TITLE
Add newrelic_synthetics_monitor and script resources

### LINKED ISSUE #65: Support for CRUD synthetics alerts and scripts
Hi! Maintaining our synthetics scripts and alerts via terraform has been a desire of ours for a while. We're looking to contribute in the coming weeks, but wanted to put this on the board and double-check whether this was already being planned or worked on.

Thanks for maintaining!

### PR DESCRIPTION
**Pull request by kyeah**
## What

This adds two new resources for maintaining synthetics monitor scripts:

```
newrelic_synthetics_monitor
newrelic_synthetics_monitor_script
```

Closes #65.

## Notes

Things I left out of this PR:
- Some validation around required/unsupported options based on the monitor type
- Monitor script locations (describing HMACs for secure locations)

And some quirks:
- I defined the script separately from the monitor since it uses a different endpoint, but New Relic doesn't support deleting a script from a monitor, so instead I just set it to `" "`. The script also doesn't have a real ID, so I'm relying on the monitor ID.

Happy to hear of better ways to do ^ or to improve the proposed API in general.

## Testing

I was able to:
- import existing SCRIPT_API monitors into the state without tfstate discrepancies
- create/modify/delete valid monitors and scripts from scratch without tfstate discrepancies

### CODE REVIEW

**Comment by kyeah** [comment]
@bridgetlane Hi! Thanks for looking at this — will this be merged and added as part of an upcoming release? Just looking to get a sense of the timeline.

**Comment by bharathkumarna** [comment]
@bridgetlane when the next release would happen

**Comment by sgsiebers** [comment]
Any updates on when this PR will get merged and released?  Would love to have these new resources available for all to use.

**Comment by paultyng** [comment]
I updated this PR and rewrote some of the commit history to clean that up, when I try to run the acceptance tests (using `TESTARGS='-cover -run TestAccNewRelicSyntheticsMonitor' make testacc`) I get:

```shell
$ TESTARGS='-cover -run TestAccNewRelicSyntheticsMonitor' make testacc
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test $(go list ./... |grep -v 'vendor') -v -cover -run TestAccNewRelicSyntheticsMonitor -timeout 120m
?   	github.com/terraform-providers/terraform-provider-newrelic	[no test files]
=== RUN   TestAccNewRelicSyntheticsMonitorScript_Basic
--- FAIL: TestAccNewRelicSyntheticsMonitorScript_Basic (2.14s)
    testing.go:538: Step 0 error: Error applying: 1 error occurred:
        	* newrelic_synthetics_monitor.foo: 1 error occurred:
        	* newrelic_synthetics_monitor.foo: error: invalid response from CreateMonitor with code 401. Message: {"code":401,"message":"HTTP 401 Unauthorized"}
        
        
        
        
=== RUN   TestAccNewRelicSyntheticsMonitor_Basic
--- FAIL: TestAccNewRelicSyntheticsMonitor_Basic (1.38s)
    testing.go:538: Step 0 error: Error applying: 1 error occurred:
        	* newrelic_synthetics_monitor.foo: 1 error occurred:
        	* newrelic_synthetics_monitor.foo: error: invalid response from CreateMonitor with code 401. Message: {"code":401,"message":"HTTP 401 Unauthorized"}
        
        
        
        
    testing.go:599: Error destroying resource! WARNING: Dangling resources
        may exist. The full state and error is shown below.
        
        Error: config is invalid: newrelic_synthetics_monitor.foo: expected frequency to be one of [1 5 10 15 30 60 360 720 1440], got 2
        
        State: <nil>
FAIL
coverage: 9.1% of statements
FAIL	github.com/terraform-providers/terraform-provider-newrelic/newrelic	3.531s
GNUmakefile:17: recipe for target 'testacc' failed
make: *** [testacc] Error 1
```

The unauthorized is possibly a problem with my account, but the other error looks some something else wrong in the test code.

**Comment by paultyng** [comment]
If someone has an account that is properly configured, can you please run these tests and post the results? Or let me know what I can do on our account to get it to work?

File: newrelic/resource_newrelic_synthetics_monitor_test.go [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+`, rName)
+}
+
+func testAccCheckNewRelicSyntheticsMonitorConfigUpdated(rName string) string {
+	return fmt.Sprintf(`
+
+resource "newrelic_synthetics_monitor" "foo" {
+  name = "%[1]s-updated"
+  type = "SIMPLE"
+  frequency = 2
```
Comment (sgsiebers) [inline newrelic/resource_newrelic_synthetics_monitor_test.go]: 2 is not an allowed monitor frequency per New Relic API: https://docs.newrelic.com/docs/apis/synthetics-rest-api/monitor-examples/manage-synthetics-monitors-rest-api#simple-monitor-calls

This is breaking your acceptance tests

**Comment by kyeah** [comment]
Thanks for the updates @paultyng / @sgsiebers. With regards to the unauthorized issue, are you using an admin API key on an account subscribed to NR Synthetics? The synthetics support is separate from things like APM.

I ran the acceptance tests locally after fixing two things:
* Updated script frequency from 2 to 5 as mentioned by @sgsiebers.
* [Updated the newrelic_synthetics_monitor type from SIMPLE to SCRIPT_BROWSER](https://github.com/terraform-providers/terraform-provider-newrelic/pull/67/files#diff-797581960ee27f1ff7ee9db7ecf9ef51R88) because SIMPLE doesn't support scripts.

```
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test $(go list ./... |grep -v 'vendor') -v -cover -run TestAccNewRelicSyntheticsMonitor -timeout 120m
?   	github.com/terraform-providers/terraform-provider-newrelic	[no test files]
=== RUN   TestAccNewRelicSyntheticsMonitorScript_Basic
--- PASS: TestAccNewRelicSyntheticsMonitorScript_Basic (7.46s)
=== RUN   TestAccNewRelicSyntheticsMonitor_Basic
--- FAIL: TestAccNewRelicSyntheticsMonitor_Basic (3.19s)
	testing.go:538: Step 0 error: Check failed: Check 8/8 error: newrelic_synthetics_monitor.foo: Attribute 'locations.0' not found
FAIL
coverage: 14.8% of statements
FAIL	github.com/terraform-providers/terraform-provider-newrelic/newrelic	10.709s
make: *** [testacc] Error 1
```

I notice when testing against `locations` it actually keys on the generated ID `locations.<id>` instead of the string index `locations.0`. It looks like most terraform provider tests that need to do this just test `field.#` instead of the actual values, so I'm guessing this check should be removed.

**Comment by sgsiebers** [comment]
I have access to a large enterprise New Relic account with synthetics licensing and have not had a problem with authorization running acceptance tests.  I posted an issue I ran into here that your code hopefully fixes:

https://github.com/terraform-providers/terraform-provider-newrelic/pull/70#issuecomment-461965484

I think it's a problem in master with the original tests trying to import a synthetics data resource that does not exist (and can't be created by the test until your code is merged).  Hopefully your test fixes that issue.  I don't have commit privs, but if you force push your latest, I'm happy to pull it down locally and confirm that tests are passing on our account if that helps get this merged.

**Comment by kyeah** [comment]
Pushed the listed changes up; let me know if the tests pass on your account. Thank you!

**Comment by sgsiebers** [comment]
@kyeah, I locally merged your branch on top of latest master and the acceptance tests work fine in my New Relic account.  The one remaining test failure also exists in master and as I indicated above appears to be a latent issue with the original synthetics data resource, not introduced here.  I suppose it's up to @paultyng whether to fix that here in this PR since this feature is related, or go ahead and merge knowing the issue already exists in master.  I'm not a maintainer so can't make that call, but I'm very interested in seeing this feature released, so happy to help out however I can.

```
$ make testacc
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test $(go list ./... |grep -v 'vendor') -v  -timeout 120m
?   	github.com/terraform-providers/terraform-provider-newrelic	[no test files]
=== RUN   TestAccNewRelicAlertChannelDataSource_Basic
--- PASS: TestAccNewRelicAlertChannelDataSource_Basic (3.35s)
=== RUN   TestAccNewRelicAlertPolicyDataSource_Basic
--- PASS: TestAccNewRelicAlertPolicyDataSource_Basic (2.08s)
=== RUN   TestAccNewRelicApplication_Basic
--- PASS: TestAccNewRelicApplication_Basic (9.91s)
=== RUN   TestParseIDs_Basic
--- PASS: TestParseIDs_Basic (0.00s)
=== RUN   TestSerializeIDs_Basic
--- PASS: TestSerializeIDs_Basic (0.00s)
=== RUN   TestProvider
--- PASS: TestProvider (0.00s)
=== RUN   TestProviderImpl
--- PASS: TestProviderImpl (0.00s)
=== RUN   TestAccNewRelicAlertChannel_Basic
--- PASS: TestAccNewRelicAlertChannel_Basic (2.62s)
=== RUN   TestAccNewRelicAlertChannel_import
--- PASS: TestAccNewRelicAlertChannel_import (1.63s)
=== RUN   TestAccNewRelicAlertCondition_Basic
--- PASS: TestAccNewRelicAlertCondition_Basic (18.28s)
=== RUN   TestAccNewRelicAlertCondition_ZeroThreshold
--- PASS: TestAccNewRelicAlertCondition_ZeroThreshold (12.64s)
=== RUN   TestAccNewRelicAlertCondition_import
--- PASS: TestAccNewRelicAlertCondition_import (8.11s)
=== RUN   TestAccNewRelicAlertCondition_nameGreaterThan64Char
--- PASS: TestAccNewRelicAlertCondition_nameGreaterThan64Char (0.02s)
=== RUN   TestAccNewRelicAlertPolicyChannel_Basic
--- PASS: TestAccNewRelicAlertPolicyChannel_Basic (5.44s)
=== RUN   TestAccNewRelicAlertPolicy_Basic
--- PASS: TestAccNewRelicAlertPolicy_Basic (2.49s)
=== RUN   TestAccNewRelicAlertPolicy_import
--- PASS: TestAccNewRelicAlertPolicy_import (2.05s)
=== RUN   TestAccNewRelicDashboard_Basic
--- PASS: TestAccNewRelicDashboard_Basic (5.25s)
=== RUN   TestAccNewRelicDashboard_import
--- PASS: TestAccNewRelicDashboard_import (1.17s)
=== RUN   TestAccNewRelicInfraAlertCondition_Basic
--- PASS: TestAccNewRelicInfraAlertCondition_Basic (7.50s)
=== RUN   TestAccNewRelicInfraAlertCondition_Where
--- PASS: TestAccNewRelicInfraAlertCondition_Where (3.08s)
=== RUN   TestAccNewRelicInfraAlertCondition_IntegrationProvider
--- SKIP: TestAccNewRelicInfraAlertCondition_IntegrationProvider (0.00s)
    resource_newrelic_infra_alert_condition_test.go:78: Environment variable ENABLE_NEWRELIC_INTEGRATION_PROVIDER is not set
=== RUN   TestAccNewRelicInfraAlertCondition_Thresholds
--- PASS: TestAccNewRelicInfraAlertCondition_Thresholds (4.86s)
=== RUN   TestAccNewRelicNrqlAlertCondition_Basic
--- PASS: TestAccNewRelicNrqlAlertCondition_Basic (3.98s)
=== RUN   TestAccNewRelicSyntheticsAlertCondition_Basic
--- FAIL: TestAccNewRelicSyntheticsAlertCondition_Basic (2.45s)
    testing.go:538: Step 0 error: Error refreshing: 1 error occurred:
        	* data.newrelic_synthetics_monitor.bar: 1 error occurred:
        	* data.newrelic_synthetics_monitor.bar: data.newrelic_synthetics_monitor.bar: The name 'tf_test_wxomop8owh' does not match any New Relic monitors.
        
        
        
        
=== RUN   TestAccNewRelicSyntheticsMonitorScript_Basic
--- PASS: TestAccNewRelicSyntheticsMonitorScript_Basic (4.66s)
=== RUN   TestAccNewRelicSyntheticsMonitor_Basic
--- PASS: TestAccNewRelicSyntheticsMonitor_Basic (3.22s)
=== RUN   TestValidationIntInInSlice
--- PASS: TestValidationIntInInSlice (0.00s)
=== RUN   TestValidationFloat64Gte
--- PASS: TestValidationFloat64Gte (0.00s)
FAIL
FAIL	github.com/terraform-providers/terraform-provider-newrelic/newrelic	104.808s
make: *** [testacc] Error 1
```

**Comment by sgsiebers** [comment]
@kyeah, I have a fix for the synthetics alert condition acceptance test failure.  If you can add me as a collaborator to your fork, I'll push my changes.  That should get all the tests passing 100%, and hopefully get this ready for @paultyng to merge.

**Comment by kyeah** [comment]
Thanks @sgsiebers! I added you as a collaborator.

**Comment by sgsiebers** [comment]
@kyeah, I'm not able to push changes to your fork yet, and don't see your repo listed in my profile.  Can you double check that I'm listed as a collaborator in your forked repo?

**Comment by kyeah** [comment]
@sgsiebers Can you check your email or [this page](https://github.com/kyeah/terraform-provider-newrelic/invitations) and see if there's an invitation waiting for you?

**Comment by sgsiebers** [comment]
Got it thanks...must have missed the collaboration invite in my email.  I pushed the acceptance test changes

**Comment by sgsiebers** [comment]
FYI...all synthetics related tests are passing now.  There are still some acceptance test failures for other resources, but those seem to be failing in master as well.  @paultyng, anything else you need to get this merged?

**Comment by paultyng** [comment]
Sorry for the delay here, things are taking a little longer as we try to ship 0.12. Will take a look at it probably next week, but just at a glance looks like its all covered, thanks!

**Comment by sgsiebers** [comment]
Awesome, thanks for the update @paultyng!

**Comment by AshleyPoole** [comment]
This would be awesome to be merged and released as trying to setup Synthetics monitoring at the moment @paultyng... Thanks for the hard work @kyeah!

**Comment by bharathkumarna** [comment]
@paultyng it would be great to merge this soon and any estimates on the next release date

**Comment by AshleyPoole** [comment]
Is anyone else able to merge this in? @appilon maybe? :-)

**Comment by AshleyPoole** [comment]
Is anyone else able to merge this in? @appilon @radeksimko

**Comment by sgsiebers** [comment]
As @paultyng indicated above, and from what I can see from other code merges, it seems everyone is very preoccupied trying to get Terraform 12 out the door, so I'm guessing that's the blocker.  In the meantime, I'm using my own fork with this code merged in and it works beautifully.

**Comment by AshleyPoole** [comment]
@sgsiebers what's involved with that on Windows?

**Comment by paultyng** [comment]
This should be going out in 1.5.0 shortly. This release also has some 0.12 support in it, so please open issues if you encounter weirdness.

**Pull request merged**